example-nextjs/utils/index.ts

```diff
@@ -0,0 +1,22 @@
+export interface CustomKeyBinding {
+  label: string
+  keybinding: any
+  callback: () => void
+  editorRef: any
+}
+
+export const addKeyBinding = ({
+  label,
+  keybinding,
+  callback,
+  editorRef,
+}: CustomKeyBinding) => {
+  editorRef?.current?.addAction({
+    id: 'label',
```

Comment:
```suggestion
    id: label,
```

Should this be ^ ? It probably doesn't make a difference while this is our only action but I'm guessing id has to be unique if we have multiple in future.

Reply:
Yeah sorry, I missed it.
Will push the fix for it.

Reply:
> I'm not really experienced enough to comment on the specific `useRef` code used here/the implementation, but as far as I can tell this now works correctly for all languages with cmd-enter on macOS.

And for this, I looked online and if we want to access the `editor` object in React this is the way to do it.
Here's the official doc [reference](https://www.npmjs.com/package/@monaco-editor/react#editor-instance)

Reply:
Hmm it looks like we might be able to make use of the `useMonaco` hook that you're already using to avoid adding a ref, judging by [these docs](https://www.npmjs.com/package/@monaco-editor/react#usemonaco)

I think you may be able to remove the `editorRef` and `handleOnMount` and add a hook like this:

```typescript
const monaco = useMonaco() // you already have this

useEffect(() => {
  if(monaco) {
      const runCodeBinding: CustomKeyBinding = {
        label: 'run',
        keybinding: CtrlEnter,
        callback: async () => runCode(inputCode, pyodide),
        monaco,
    }
    addKeyBinding(runCodeBinding)
  }
}, [monaco])
```

The docs suggest that this is an alternative to using the ref and I think it's probably clearer. The way it should work from my understanding of those docs is that `monaco` is set asynchronously when the editor loads, and at that point the `useEffect` hook should be triggered because of that dependency.

Reply:
If you see the definition of method `addKeyBinding`.
```typescript
export const addKeyBinding = ({
  label,
  keybinding,
  callback,
  editorRef,
}: CustomKeyBinding) => {
  editorRef?.current?.addAction({
    id: label,
    label,
    keybindings: [keybinding],
    precondition:
      '!suggestWidgetVisible && !markersNavigationVisible && !findWidgetVisible',
    run: callback,
  })
}
```

I need access of `editor` object to add the action to it.
By your suggestion now I am accessing `monaco.editor.addAction` which is undefined.
I also tried to check if `monaco.editor.addCommand` is there but it's not.
But when I use `editorRef` and accessing `editorRef.current.addAction` it is there.
And as show about I am able to add the action cmd+enter and ctrl+enter

Please suggest if you can find something else.  

One thing which is avoidable is use of `useMonaco` because I will get access to moanco in the `handleEditorDidMount` second param and from there I can get the keybinding value and using a state I can achieve the same.

Reply:
Yep sorry, see my edit! Tried to make an example and hit the same blocker. Seems `monaco.editor` != the editor instance, which is pretty confusing.

Good call about the `useMonaco` hook, agreed that we can remove that and get the monaco instance from `handleEditorDidMount`instead. Other than that though I think this is the approach we need to use, and your code looks good to me.

Reply:
Cool,
Let me change the code to get `monaco` object from `handleEditorDidMount`

Reply:
@mcintyre94 
So this is how it will look now.
What do you think?

```typescript
  const [monaco, setMonaco] = useState<Monaco>(null)

  useEffect(() => {
    if (monaco && inputCode && pyodide) {
      const runCodeBinding: CustomKeyBinding = {
        label: 'run',
        keybinding: monaco?.KeyMod?.CtrlCmd | monaco?.KeyCode?.Enter,
        callback: async () => runCode(inputCode, pyodide),
        editorRef,
      }
      addKeyBinding(runCodeBinding)
    }
  }, [monaco, inputCode, pyodide])

  function handleEditorDidMount(editor, monaco) {
    editorRef.current = editor
    setMonaco(monaco)
  }

```

Reply:
I want to avoid invokation of `addKeyBinding` when `monaco` and `pyodide` are null

Reply:
I think this is looking good!

I find it pretty frustrating that `monaco.editor` and `editor` are different objects. It seems `monaco.editor` is mainly set for the configuration of editors, whereas `editor` just refers to the actual instance. The [type definition](https://github.com/suren-atoyan/monaco-react/blob/dd9c771a9e176c4b255e620708bcd5d8ccda251a/src/types.d.ts#L15) of `onMount` is also not very helpful. FWIW, `CodeMirror.editor` and the `editor` exposed by the API are the same.

Looks like this might be the best way forward given we need to account for the loading of `pyodide` and the editor. I also tested `Cmd+Enter` and `Ctrl+Enter` and both work well with the current implementation!

Reply:
Yeah, I was so irritated by this. I looked into types but no help for the `onMount` return type.

`CodeMirror` is something else right [this](https://codemirror.net/) is the one you are mentioning about?

Reply:
This looks good to me! 👍

Reply:
@mcintyre94  @jacobjzhang 

I am thinking we should keep [this](https://github.com/slipHQ/run-wasm/blob/82b2a064b2e8cb2ca9512a6d58835a75adf762af/example-nextjs/pages/index.tsx#L37) functions and other definitions below `useEffect`.
in pages all of the three pages.

What do you suggest? Should I refactor it in this PR only or it will be better to in another?
As usually I see people suggesting `useEffect` should be above

Reply:
I'm not familiar with that advice but if you want to/need to do it then I'd just update in this PR. It'd be confusing to review a separate PR making that change without all the context of this one and won't touch any more files/code than this PR already does.